HazelcastInstanceNotActiveException after configuration refresh (5.0.x)

Hi,
I noticed an issue with the hazelcast ticket registry: 
During a configuration refresh (for example from the dashboard `refresh` button), the Hazelcast instance is shut down and not reactivated afterward.
Any subsecant use of the hazelcast instance will then result in a `com.hazelcast.core.HazelcastInstanceNotActiveException` being thrown.

If I understood correctly the behaviour of the different anotations, the function [`org.apereo.cas.config.HazelcastInstanceConfiguration.hazelcast()`](https://github.com/apereo/cas/blob/5.0.x/support/cas-server-support-hazelcast-ticket-registry/src/main/java/org/apereo/cas/config/HazelcastInstanceConfiguration.java#L64) is a Bean and is thus only run once to generate an `HazelcastInstance` but [`org.apereo.cas.config.HazelcastInstanceConfiguration.hazelcastTicketRegistry()`](https://github.com/apereo/cas/blob/5.0.x/support/cas-server-support-hazelcast-ticket-registry/src/main/java/org/apereo/cas/config/HazelcastInstanceConfiguration.java#L53) which is marked as `RefreshScope` will be destroyed (which will [shutdown](https://github.com/apereo/cas/blob/5.0.x/support/cas-server-support-hazelcast-ticket-registry/src/main/java/org/apereo/cas/ticket/registry/HazelcastTicketRegistry.java#L116) the `HazelcastInstance`) and recreated (with the same inactive `HazelcastInstance`).

I don't realy know what is the proper solution to fix this issue:
- Mark the `hazelcast()` bean as `RefreshScope`, but this doesn't work because `com.hazelcast.instance.HazelcastInstanceProxy` is final,
- Don't mark `hazelcast()` as Bean, I've tested it, it seems to work but there might be a reason why it was a Bean in the first place. And it might cause some trouble/duplicate instance if the function hazelcast is used anywhere else in the code except in hazelcastTicketRegistry.
- Don't shutdown the `HazelcastInstance` on `HazelcastTicketRegistry` destroy, but shut it down somewhere else (but where?)

Additional informations

- [x] CAS server version: **5.0.4-SNAPSHOT**
- [x] Container/environment type/version (Tomcat, Jetty, Java, Linux/Windows, etc): **Tomcat**
- [x] Steps to duplicate the problem along with your CAS settings and properties:

1. Go to https://localhost:8443/status/dashboard
2. Press the 'refresh' button and confirm
3. Go to https://localhost:8443/status

- [x] Stacktraces and CAS logs where appropriate.

```
2017-04-21 16:46:03,602 INFO [org.apereo.cas.configuration.CasConfigurationRebinder] - <Reloaded CAS configuration cas-org.apereo.cas.configuration.CasConfigurationProperties>
2017-04-21 16:46:04,304 INFO [org.apereo.cas.ticket.registry.HazelcastTicketRegistry] - <Shutting down Hazelcast instance localhost>
2017-04-21 16:46:04,352 WARN [org.springframework.beans.factory.support.DisposableBeanAdapter] - <Invocation of destroy method 'close' failed on bean with name 'scopedTarget.hazelcastTicketRegistry': com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!>
2017-04-21 16:46:31,188 INFO [org.apereo.cas.services.DefaultServicesManagerImpl] - <Loaded 2 services from InMemoryServiceRegistryDaoImpl.>
2017-04-21 16:46:31,224 ERROR [org.apereo.cas.ticket.registry.DefaultTicketRegistryCleaner] - <Error creating bean with name 'scopedTarget.hazelcastTicketRegistry' defined in class path resource [org/apereo/cas/config/HazelcastInstanceConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.apereo.cas.ticket.registry.TicketRegistry]: Factory method 'hazelcastTicketRegistry' threw exception; nested exception is com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!>
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'scopedTarget.hazelcastTicketRegistry' defined in class path resource [org/apereo/cas/config/HazelcastInstanceConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.apereo.cas.ticket.registry.TicketRegistry]: Factory method 'hazelcastTicketRegistry' threw exception; nested exception is com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:599) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1128) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1022) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:512) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory$2.getObject(AbstractBeanFactory.java:345) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.cloud.context.scope.GenericScope$BeanLifecycleWrapper.getBean(GenericScope.java:359) ~[spring-cloud-context-1.1.6.RELEASE.jar!/:1.1.6.RELEASE]
	at org.springframework.cloud.context.scope.GenericScope.get(GenericScope.java:176) ~[spring-cloud-context-1.1.6.RELEASE.jar!/:1.1.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:340) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.aop.target.SimpleBeanTargetSource.getTarget(SimpleBeanTargetSource.java:35) ~[spring-aop-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:192) ~[spring-aop-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at com.sun.proxy.$Proxy83.getTickets(Unknown Source) ~[?:?]
	at org.apereo.cas.ticket.registry.DefaultTicketRegistryCleaner.clean(DefaultTicketRegistryCleaner.java:63) ~[cas-server-core-tickets-5.0.4.jar!/:5.0.4]
	at org.apereo.cas.ticket.registry.DefaultTicketRegistryCleaner$$FastClassBySpringCGLIB$$29f046b2.invoke(<generated>) ~[cas-server-core-tickets-5.0.4.jar!/:5.0.4]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204) ~[spring-core-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:720) ~[spring-aop-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157) ~[spring-aop-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99) ~[spring-tx-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:282) ~[spring-tx-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96) ~[spring-tx-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:655) ~[spring-aop-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.apereo.cas.ticket.registry.DefaultTicketRegistryCleaner$$EnhancerBySpringCGLIB$$bda1d4c8.clean(<generated>) ~[cas-server-core-tickets-5.0.4.jar!/:5.0.4]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_121]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_121]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_121]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_121]
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:65) ~[spring-context-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54) ~[spring-context-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_121]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) ~[?:1.8.0_121]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_121]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) ~[?:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.apereo.cas.ticket.registry.TicketRegistry]: Factory method 'hazelcastTicketRegistry' threw exception; nested exception is com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	... 36 more
Caused by: com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!
	at com.hazelcast.instance.HazelcastInstanceProxy.getOriginal(HazelcastInstanceProxy.java:288) ~[hazelcast-3.7.2.jar!/:3.7.2]
	at com.hazelcast.instance.HazelcastInstanceProxy.getMap(HazelcastInstanceProxy.java:92) ~[hazelcast-3.7.2.jar!/:3.7.2]
	at org.apereo.cas.ticket.registry.HazelcastTicketRegistry.<init>(HazelcastTicketRegistry.java:55) ~[cas-server-support-hazelcast-ticket-registry-5.0.5-SNAPSHOT.jar!/:5.0.4]
	at org.apereo.cas.config.HazelcastInstanceConfiguration.hazelcastTicketRegistry(HazelcastInstanceConfiguration.java:58) ~[cas-server-support-hazelcast-ticket-registry-5.0.5-SNAPSHOT.jar!/:5.0.4]
	at org.apereo.cas.config.HazelcastInstanceConfiguration$$EnhancerBySpringCGLIB$$2cce8644.CGLIB$hazelcastTicketRegistry$1(<generated>) ~[cas-server-support-hazelcast-ticket-registry-5.0.5-SNAPSHOT.jar!/:5.0.4]
	at org.apereo.cas.config.HazelcastInstanceConfiguration$$EnhancerBySpringCGLIB$$2cce8644$$FastClassBySpringCGLIB$$abc8e13b.invoke(<generated>) ~[cas-server-support-hazelcast-ticket-registry-5.0.5-SNAPSHOT.jar!/:5.0.4]
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228) ~[spring-core-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:356) ~[spring-context-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.apereo.cas.config.HazelcastInstanceConfiguration$$EnhancerBySpringCGLIB$$2cce8644.hazelcastTicketRegistry(<generated>) ~[cas-server-support-hazelcast-ticket-registry-5.0.5-SNAPSHOT.jar!/:5.0.4]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_121]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_121]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_121]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_121]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588) ~[spring-beans-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
	... 36 more

```

- [ ] Configuration/Overlay pushed to https://github.com/cas-projects/cas-issues/

--- a/support/cas-server-support-hazelcast-ticket-registry/src/main/java/org/apereo/cas/config/HazelcastInstanceConfiguration.java
+++ b/support/cas-server-support-hazelcast-ticket-registry/src/main/java/org/apereo/cas/config/HazelcastInstanceConfiguration.java
@@ -51,7 +51,6 @@ public class HazelcastInstanceConfiguration {
     private CasConfigurationProperties casProperties;
         
     @Bean(name = {"hazelcastTicketRegistry", "ticketRegistry"})
-    @RefreshScope
     public TicketRegistry hazelcastTicketRegistry() {
         final HazelcastTicketRegistry r = new HazelcastTicketRegistry(hazelcast(),
                 casProperties.getTicket().getRegistry().getHazelcast().getMapName(),

